src/middlewares/Cache.ts

```diff
@@ -0,0 +1,151 @@
+        /* istanbul ignore else */
+        if (!entry) { return next(options).then((response) => this.setCache(key, response)); }
+
+        try {
+            const cachedEntry: ICacheEntry = JSON.parse(entry);
+
+            // If expired, pass to next and set cache again
+            if (options.cache.maxAge && +Date.now() > (cachedEntry.timestamp + options.cache.maxAge * 1000)) {
+                return next(options).then((response) => this.setCache(key, response));
+            }
+
+            return this.getCachedResponse(cachedEntry);
+        } catch (e) { // Failed serving from cache, proceed normally
+            /* istanbul ignore next */
+            return next(options).then((response) => this.setCache(key, response));
+        }
+    }
+
+    /**
+     * Persists the cache to storage configured, or in memory if that fails.
+     *
+     * @param   {string}        key
+     * @param   {FetchResponse} response
+     * @returns {FetchResponse}
+     */
+    private setCache(key: string, response: FetchResponse<any>): FetchResponse<any> {
+        const headers: FetchHeaders = {};
+        const clone: FetchResponse<any> = response.clone();
+
+        /* istanbul ignore else */
+        if (clone.headers) { clone.headers.forEach((value: string, name: string) => headers[name] = value ); }
+
+        clone.text().then((value: string) => {
+            const entry = JSON.stringify({
+                value,
+                type: clone.type,
+                url: clone.url,
+                status: clone.status,
+                statusText: clone.statusText,
+                timestamp: +Date.now(),
+                headers
+            });
+
+            if (this.storage) {
+                try {
+                    this.storage.setItem(key, entry);
+                } catch (e) {
+                    /* istanbul ignore next */
+                    this.fallbackStorage[key] = entry;
+                }
+            } else {
+                this.fallbackStorage[key] = entry;
+            }
+        });
+
+        return response;
+    }
+
+    /**
+     * Returns a Promise that resolves to Response object using cached value if available.
+     * Sets the cache and returns a promise that resolves to Response if the cache is non existent or expired.
+     *
+     * @return {Promise}
+     */
+    private getCachedResponse(cachedEntry: ICacheEntry): Promise<FetchResponse<any>> {
+        const response: FetchResponse<any> = new Response(cachedEntry.value, cachedEntry);
+
+        response.cacheUsed = true;
+        response.cacheTimestamp = cachedEntry.timestamp;
+
+        return Promise.resolve(response);
+    }
+
+    /**
+     * Gets the normalized cache key which is unique to request with respect to uri and params.
+     *
+     * @param  {FetchRequest} options
+     * @return {string}
+     */
+    private getCacheKey(options: FetchRequest): string {
+        let key: string = (options.cache && options.cache.key) || options.url || "";
```

Comment:
we have optional string typedef `?: string` and thus build fails without coalesce `|| ""`